### ISSUE TITLE
Please support "editor.codeActionsOnSave"

### ISSUE THREAD

**Issue by ahuglajbclajep** [Tom-Bonnike/vscode-formatting-toggle#54]
`editor.formatOnSave` as well as [editor.codeActionsOnSave](https://code.visualstudio.com/updates/v1_31#_fix-all-source-actions) also edits the code on save.
Some linters and formatters, such as [eslint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint#release-notes), use `editor.codeActionsOnSave`, so it's useful to be able to control this setting as well.

**Comment by ahuglajbclajep** [comment]
The default value of `editor.codeActionsOnSave` is `{}` and also, to re-enable it, you need to remember the original settings, so you can simply extend the current implementation with it may not be enough to do so.
https://github.com/Tom-Bonnike/vscode-formatting-toggle/blob/2.0.0/src/registerCommand.ts#L22-L30

**Comment by Tom-Bonnike** [comment]
> so you can simply extend the current implementation with it may not be enough to do so.

PRs welcome!

**Comment by supertry55** [comment]
#54

**Comment by yancyknight** [comment]
@Tom-Bonnike I'd like to take a shot at this. I think a good approach would be to look for a special character, such as `@` or `#`, to indicate a setting in an object.
```json
{
  "formattingToggle.affects": [
    "editor.codeActionsOnSave@source.fixAll.eslint",
    "editor.codeActionsOnSave#source.fixAll.eslint"
  ]
}
```
In the loop currently updating the settings we would check for that special character and if it's there we do this:
```javascript
const [ outerSetting, innerSetting ] = setting.split(specialCharacter);
editorConfiguration.update(
  outerSetting,
  { [innerSetting]: !isFormattingActivated },
  ConfigurationTarget.Global
)
```

I can think of other approaches as well if you don't like this one. We could allow objects in `formattingToggle.affects` or we could have a separate `formattingToggle.affects.codeActionsOnSave` setting for instance.

Like I said, I'm happy to submit a PR, I just want to make sure you're on board before I do the work.

**Comment by Tom-Bonnike** [comment]
👋 Hi Yancy! Thanks for offering your help! 🌟 

---

> I think a good approach would be to look for a special character, such as @ or #, to indicate a setting in an object.

Mhm I see, that’s a good workaround but in an ideal world we’d manage not to introduce such a quirk 🤔 

For every `WorkspaceConfiguration` method, including `update`, the [VSCode API documentation](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration) states:

Parameter | Description
-- | --
section: string | Configuration name, supports dotted names.

Somehow it works fine for `get` but not `update`, it would have been amazing if they had handled it based on [what the configuration expects](https://code.visualstudio.com/api/references/contribution-points#contributes.configuration), then it would actually already be working 😄 

Maybe through `WorkspaceConfiguration#inspect` or `WorkspaceConfiguration#has` we could somehow be smart about it and traverse the path until we find a boolean setting? Seems complex 🤔 

---

> We could allow objects in formattingToggle.affects

Yeah, I think that might be the way to go?

```json
{
  "formattingToggle.affects": [
    // String
    "editor.formatOnPaste",

    // Array
    ["editor.codeActionsOnSave", "source.fixAll.eslint"],

    // Object
    { "scope": "editor.codeActionsOnSave", "section": "source.fixAll.eslint" },
  ]
}
```

I guess the array solution might be the best one because it could work for deeply nested settings?

```json
{
  "formattingToggle.affects": [
    ["editor.codeActionsOnSave", "foo.bar", "baz"],
  ]
}
```


```ts
affectsConfiguration.forEach(setting => {
  if (!Array.isArray(setting)) {
    return configuration.update(
      setting,
      !isFormattingActivated,
      ConfigurationTarget.Global,
    )
  }
  
  const section = setting.at(-1)
  const scope = workspace.joinPath('', setting.slice(0, -1))

  workspace.getConfiguration(section, scope).update(
    setting,
    !isFormattingActivated,
    ConfigurationTarget.Global,
  )
})
```

We’d need to document it but, I think it might be more natural than having to learn about a special character?

> or we could have a separate formattingToggle.affects.codeActionsOnSave setting for instance

… or add a super specific condition like `if setting.includes('editor.codeActionsOnSave')`… after all this extension was created with formatting settings in mind as indicated in the README: 
> 💡 Formatting Toggle was created with formatting settings in mind but allows you to toggle any boolean setting that lives at the root of the VSCode configuration.

**Comment by yancyknight** [comment]
Thanks for the quick and thorough reply!

If you want to support any deeply nested property I agree that the array/tuple syntax you proposed is the nicest. That said, I got here just looking for `editor.codeActionsOnSave` and I don't see myself needing anything else so a hard coded `if` sounds good to me. 

I'll get a PR put together in the next couple days.

**Comment by Tom-Bonnike** [comment]
> I'll get a PR put together in the next couple days unless you change your mind before then.

Don’t think I will, you can go ahead! Thank you!

**Comment by rickstaa** [comment]
@yancyknight Great PR. Thanks a lot! Hope this gets merged soon. 🚀

**Comment by Tom-Bonnike** [comment]
Yup, thanks Yancy. I looked through it quickly as soon as you opened it but I still need to find the time to do a proper review, sorry folks 🙏

**Comment by Tom-Bonnike** [comment]
Just published 3.1.0 with support for boolean configs under `editor.codeActionsOnSave`. Thanks a lot @yancyknight! 🎉
